### ISSUE TITLE
Indirect role recursion (due to callback) + lazy evaluation in variables prevents using vars.

### ISSUE THREAD

**Issue by nhosoi** [linux-system-roles/logging#102]
**Problem description:**
In the top level task, set internal variables to the corresponding external ones to pass them to the included role as follows, in the top level default file, the default value is set to external_var.
```
vars:
  __internal_var: "{{ exteranl_var }}"
include_role:
  name: "{{ role_path }}/roles/rsyslog"
```
Then, in the role, it includes another level of role as follows.  Please note that the deploy.yml is not in the top level.
```
  include_role:
    name: "{{ role_path }}/../../../"
    tasks_from: deploy.yml
```
In deploy.yml, __internal_var is referred.  Due to the ansible's lazy variable evaluation, external_var is accessed at the first time, then it issues an error - the external_var variable is not defined when the external_var is not set in the inventory file.  If it is set in the inventory file, the value is correctly evaluated and used.

For more details, see also the conversation starting with https://github.com/linux-system-roles/logging/pull/99#discussion_r400578960

**Comment by pcahyna** [comment]
Can you please show the default file for this example?

**Comment by pcahyna** [comment]
> Then, in the role, it includes another level of role as follows. Please note that the deploy.yml is not in the top level.

Where is it then? At what level?

**Comment by nhosoi** [comment]
> Can you please show the defaults file for this example?

Top level defaults file?  If yes, this is the one.
https://github.com/nhosoi/logging/blob/rsyslog_system_log_dir/defaults/main.yml

**Comment by nhosoi** [comment]
> > Then, in the role, it includes another level of role as follows. Please note that the deploy.yml is not in the top level.
> 
> Where is it then? At what level?

deploy.yml is here:
`logging/roles/rsyslog/tasks/deploy.yml`
The top level defaults & tasks are here:
`logging/{defaults,tasks}/main.yml`
The idea is the top level provides a provider independent interface and providers are listed in the roles directory.  In the future, other providers will be added as follows,
```
logging/roles/fluentd/
logging/roles/loki/
logging/roles/...
```

**Comment by pcahyna** [comment]
@nhosoi 

> Then, in the role, it includes another level of role as follows. Please note that the deploy.yml is not in the top level.
> 
> ```
>   include_role:
>     name: "{{ role_path }}/../../../"
>     tasks_from: deploy.yml
> ```
> 

What role does "in the role" refer to? If it includes deploy.yml from a role a level above, and deploy.yml is in the `rsyslog` role, does it means that we are in a role below rsyslog? I.e. are there three levels involved - `logging`, `rsyslog`, and the role which contains the task above?

**Comment by pcahyna** [comment]
Hmm I think I reproduced your problem. Buit it does not seem to be related to nested roles at all. If I flatten the structure, the problem persists. The problem seems rather to be that from rsyslog you call basics and from basics you call back rsyslog. I.e. the problem is in dynamic scope nesting, not static scope nesting.

**Comment by pcahyna** [comment]
Here is a simplified demo to show the problem. https://github.com/pcahyna/nested_roles_demo

**Comment by pcahyna** [comment]
@nhosoi  I updated the demo and added a brief README. THe demo shows that nested roles are not really a problem, recursion is.

**Comment by pcahyna** [comment]
fixed by #108 ?

**Comment by nhosoi** [comment]
> fixed by #108 ?

@pcahyna, let me test it.
